lib/WUI/nhttp/segmented_json.cpp

```diff
@@ -0,0 +1,85 @@
+#include "segmented_json.h"
+#include "../json_encode.h"
+
+#include <cassert>
+#include <cstdio>
+#include <cstdarg>
+#include <cinttypes>
+
+namespace nhttp {
+
+JsonRenderer::ContentResult JsonRenderer::Output::output(size_t resume_point, const char *format, ...) {
+    va_list params;
+    va_start(params, format);
+    const size_t written = vsnprintf(reinterpret_cast<char *>(buffer), buffer_size, format, params);
+    va_end(params);
+
+    if (written < buffer_size) {
+        // Note: If the size is equal, the last byte of output was overwritten
+        // by \0. We don't _need_ that \0 there, so we would be fine without
+        // it, but there seems to be no way to tell vsnprintf to not put it there.
+        //
+        // Therefore, if the written is equal, then we _didn't_ fit.
+        buffer += written;
+        buffer_size -= written;
+        written_something = true;
+        return ContentResult::Complete;
+    } else {
+        this->resume_point = resume_point;
+        return written_something ? ContentResult::Incomplete : ContentResult::BufferTooSmall;
+    }
+}
+
+JsonRenderer::ContentResult JsonRenderer::Output::output_field_str(size_t resume_point, const char *name, const char *value) {
+    JSONIFY_STR(value);
+    return output(resume_point, "\"%s\":\"%s\"", name, value_escaped);
+}
+
+JsonRenderer::ContentResult JsonRenderer::Output::output_field_str_format(size_t resume_point, const char *name, const char *format, ...) {
+    va_list params1, params2;
+    va_start(params1, format);
+    va_copy(params2, params1);
+    // First, discover how much space we need for the formatted string.
+    char first_buffer[1];
+    // +1 for \0
+    const size_t needed = vsnprintf(first_buffer, 1, format, params1) + 1;
```

Comment:
I'm not sure if I understand the intent correctly - `first_buffer` is 1-byte long, so vsnprintf is only used to return the buffer size necessary without actually writing anything into `first_buffer`?

It looks that way since the C standard says:
_vsnprintf returns the number of characters that would have been written if n had been sufficiently large, not counting the terminating null character._

In this case of non-time critical code it is completely ok to call the relatively costly vsnprintf twice.

Reply:
Yes.